replace routerFunc parameter with entity

burrbd opened the pull request:
@buddhamagnet For discussion. This PR would require some of the yet-to-be-added processing in the RAML package. See TODOs in tests.

burrbd committed:
replace routerFunc parameter with entity

buddhamagnet committed:
adjust processing to accept verb correctly

buddhamagnet committed:
pass tests

buddhamagnet commented:
OK made some changes so the new code processes the verbs correctly and passed the tests.

I also made a change to the `raml` package so it collects all the verbs associated with an endpoint into a map and exposes that as a method on the `Resource` value meaning less repetitive code in the consuming application.

Looking good so far but I'd welcome your comments on my changes. The new consuming code I am usinh in goberry to test this now looks like this:

``` go
func routerFunc(ep *ramlapi.Endpoint) {
    route := router.
        Methods(ep.Verb).
        Path(ep.Path).
        Handler(RouteMap[ep.Handler])

    if len(ep.QueryParameters) != 0 {
        for _, query := range ep.QueryParameters {
            if query.Pattern != "" {
                route.Queries(query.Pattern, fmt.Sprintf("{%s:%s}", query.Key, query.Pattern))
            } else {
                route.Queries(query.Key, "")
            }
        }
    }
}
```

buddhamagnet committed:
remove router-specific syntax in test RAML

buddhamagnet committed:
simply code with newest update to raml package

buddhamagnet committed:
whoops

buddhamagnet commented:
@burrbd I thought the same about EndpointSet but left it, I think it's redundant, yes.

burrbd committed:
remove EndpointSet and add errors

burrbd commented:
Removed `EndpointSet`, also added some error handling, which has changed the function signature of `processResource()` and `Build()`. Not sure if this is okay. Couldn't work out why `processResource` returned string `path` anyway.

An alternative to this error handling might be to call log.Fatal() for any of these errors?

buddhamagnet commented:
Yeah I think that string return from `processResource` is a hangover from the first pass, feel free to change that too if that's true.

buddhamagnet commented:
I'd work in conjunction with this branch which contains the last commit from this PR and the latest `raml` with the new convenience methods: https://github.com/EconomistDigitalSolutions/goberry/tree/dependencies/latest-ramlapi.

Not sure about doing fatals in package code, that prevents the consuming code from deciding what to do next? But I'm half cut now so will have to have a look when I'm not in Paris and haven't consumed half a gallon of wine.

buddhamagnet commented on ramlapi.go:
```diff
@@ -1,12 +1,72 @@
+
+func appendEndpoint(s []*Endpoint, verb string, method *raml.Method) ([]*Endpoint, error) {
+	if method.DisplayName == "" {
+		return s, errors.New(`"DisplayName" property not set in RAML method`)
```
This looks a little ugly. 

``` go
return s, errors.New("DisplayName property not set in RAML method")
```

is better?

Also could we document the errors as package variables?

buddhamagnet committed:
structs first, methods next, then exported, then unexported

burrbd commented:
@buddhamagnet: **Re: GetTestEndpoint/displayName**

I was accounting for a `displayName` property like "First Name". There's no requirement for a one word `displayName`. From the RAML docs:

> (Optional) The displayName attribute specifies the parameter's display name. It is a friendly name used only for display or documentation purposes. If displayName is not specified, it defaults to the property's key (the name of the property itself).

By adding an error for missing "displayName" we're making the property a requirement for `ramlapi`.

rlhatcher commented:
We definitely need a better way to specify what we would want the handler name to be. DisplayName was an initial guess, but I agree that it doesn't fit well with the spec. The only other idea I had was to munge together the verb and the path somehow but I expect that would get pretty ugly pretty fast. The only real requirement is that we get something that will be guaranteed to be unique across a single RAML file.

buddhamagnet commented:
@burrbd @rlhatcher I guess it makes sense to both link the RAML displayName to handlers and allow multiple word displayNames, although why you would actually do that in the context of `ramlapi` I have no idea.

buddhamagnet commented:
@burrbd this change has broken `ramlgen` so needs to be adjusted before we can merge.

buddhamagnet committed:
bring ramlgen in line with ramlapi changes

burrbd commented:
@rlhatcher I think verb + path makes a lot of sense to guarantee uniqueness and ensure the developer can see which handler is which.

```
/article/{id}
PostArticleId
GetArticleId
```

Not great though. Could be confusing. Maybe suffix with "Handler"?

buddhamagnet committed:
trim down function names and init regex

rlhatcher commented:
I thought so as well except that we already have things like `/applications/{application}/objects/{object}/{applicationid}`
so calling something `GetApplicationsSalesforceObjectsCustomer` seems a bit ugly

buddhamagnet committed:
export display name normaliser and use in ramlgen

buddhamagnet commented:
@burrbd @rlhatcher I've brought `ramlgen` in line with the changes to the ramlapi changes in this branch and made some small adjustments.

I think `displayName` is fine for now, don't really like the idea of starting to mung it together from other attributes when we already have one, so I think the current state, where the edge case of having a multiple word `displayName` (that someone familiar with the API would be aware of) is also handled is fine.

I just tested all this in https://github.com/EconomistDigitalSolutions/goberry/tree/dependencies/latest-ramlapi and all looks good. I'm happy for this to be merged and this will be a major version bump as it contains API breaking changes - so any consuming application that needs to support multiple query parameters will need to upgrade. @rlhatcher you want to review as well?

buddhamagnet committed:
update changelog

buddhamagnet committed:
adjust example in README

buddhamagnet committed:
pass ramlgen test

buddhamagnet committed:
use short variable decs and extract structs out of functions

buddhamagnet committed:
comment exported functions and structs

burrbd commented:
Yeah, revisit the handler naming later.

I'm okay to merge, but my one niggle is the `verb` argument and the `Methods()` convenience method. It doesn't feel quite right to me. In the buddhamagnet/raml, what about `func (r *Resource) Methods() []*Method {` and implement stringer on the raml resource method?

Then you have:

``` go
ep := &Endpoint{
    Verb:        method.String(),
    Handler:     variableize(method.DisplayName),
    Description: method.Description,
}
```

burrbd commented:
PR here https://github.com/buddhamagnet/raml/pull/1 to support Method Names (rather than String()). Looks like the OP of the original RAML package intended to use this property for the verb name.

buddhamagnet commented:
@burrbd yep all good, merged that one.

burrbd committed:
update deps

burrbd committed:
update appendEndpoints and update dependency

buddhamagnet committed:
amend ramlgen
